src/platforms/stlinkv3/usbdfu.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Copyright (C) 2013 Gareth McMullin <gareth@blacksphere.co.nz>
+ * Portions (C) 2020-2021 Stoyan Shopov <stoyan.shopov@gmail.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <string.h>
+#include <libopencm3/cm3/systick.h>
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/stm32/gpio.h>
+#include <libopencm3/cm3/scb.h>
+#include <libopencm3/cm3/scs.h>
+
+#include "usbdfu.h"
+#include "general.h"
+
+uintptr_t app_address = APP_START;
+
+void dfu_detach(void)
+{
+	scb_reset_system();
+}
+
+int main(void)
+{
+	/* Use Top of ITCM Flash as magic marker */
+	volatile uint32_t *magic = (volatile uint32_t *)0x3ff8;
+	rcc_periph_clock_enable(RCC_GPIOA);
+	/* On the Mini, NRST is on the footprint for the 1.27 mm Jumper
+	 * to the side of the USB connector.
+	 * With debugger connected, ignore reset. Use debugger to enter!
+	 */
+	bool force_bootloader;
+	force_bootloader = (!(SCS_DHCSR & SCS_DHCSR_C_DEBUGEN) && ((RCC_CSR & RCC_CSR_RESET_FLAGS) == RCC_CSR_PINRSTF));
+	RCC_CSR |= RCC_CSR_RMVF;
+	RCC_CSR &= ~RCC_CSR_RMVF;
+	if (force_bootloader || ((magic[0] == BOOTMAGIC0) && (magic[1] == BOOTMAGIC1))) {
```

Comment:
the extra inner parens around each `magic[N]` check should be removed as redundant